app/controllers/shipit/stats_controller.rb

```diff
@@ -0,0 +1,54 @@
+module Shipit
+  class StatsController < ShipitController
+    def show
+      @stats = {
+        average: 0,
+        min: 0,
+        max: 0,
+        median: 0,
+        success: 0
+      }
+      @diffs = {}
+
+      @previous_deploys = stack.deploys.where("created_at BETWEEN ? AND ?", 14.days.ago, 7.days.ago)
+      prev_deploy_durations = @previous_deploys.map{|d| d.duration.value }
+      if prev_deploy_durations.size > 0
+        @prev_stats = {
+          count: @previous_deploys.size,
+          average: prev_deploy_durations.inject{ |sum, el| sum + el }.to_f / prev_deploy_durations.size,
+          min: prev_deploy_durations.min,
+          max: prev_deploy_durations.max,
+          median: (prev_deploy_durations.sort[(prev_deploy_durations.length - 1) / 2] + prev_deploy_durations.sort[prev_deploy_durations.length/2]) / 2.0,
+          success: (@previous_deploys.select{|d| d.success? }.size / @previous_deploys.size.to_f)*100
+        }
+      end
+
+      @deploys = stack.deploys.where("created_at > ?", 7.days.ago)
+      @deploy_durations = @deploys.map{|d| d.duration.value }
+      if @deploy_durations.size > 0
+        @stats = {
+          count: @deploys.size,
+          average: @deploy_durations.inject{ |sum, el| sum + el }.to_f / @deploy_durations.size,
+          min: @deploy_durations.min,
+          max: @deploy_durations.max,
+          median: (@deploy_durations.sort[(@deploy_durations.length - 1) / 2] + @deploy_durations.sort[@deploy_durations.length/2]) / 2.0,
```

Comment:
There seem to be some code duplication here, and the controller action is quite more complex than I'd like, as well as very procedural.

IMO it could benefit from being extracted as a model (not a db backed one, just some plain old ruby class).

Something like `DeployStats.new(deploys).count/average/etc`. It will also be much much easier to unit test.

Reply:
Agreed this is very ugly.  Lots of duplication and unnecessary complexity for a controller.